crowbar_framework/app/controllers/api/upgrade_controller.rb

```diff
@@ -64,6 +64,23 @@ def services
     end
   end
 
+  api :POST, "/api/upgrade/controllers", "Orchestrate the upgrade of controller nodes"
+  api_version "2.0"
+  def controllers
+    # First node to upgade is DRBD slave
+    # FIXME: this actually does all drbd slaves in all clusters, do we want that?
+    NodeObject.find("drbd_rsc_postgresql_master:true").each do |node|
+      node_api = Api::Node.new node.name
+      next if node_api.upgraded?
+      node_api.pre_upgrade
```

Comment:
other API calls do report the status on the execution. These functions (`pre_upgrade`, `upgrade` and `post_upgrade`) would benefit of having a return code. A boolean for success/failure may be enough.